pennylane/transforms/split_non_commuting.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the tape transform that splits non-commuting terms
+"""
+# pylint: disable=protected-access
+import pennylane as qml
+import numpy as np
+
+
```

Comment:
```suggestion

@qml.batch_transform
```
This isn't a required suggestion, but might be worth adding - it makes this transform work on both tapes _and_ qnodes :)

Reply:
> Coming back to this: Do we want `split_non_commuting()` to be a function for users in general or is it an internal tool?

It is mostly an internal tool, but I can see cases where it might come up that a user might want the control of applying this directly 🤔 But yes, you make a good point.

One example I can think of is [`hamiltonian_expand`](https://github.com/PennyLaneAI/pennylane/blob/97f0277d92adc24765d19fd49f89afb60bd6b156/pennylane/transforms/hamiltonian_expand.py#L21). This one _doesn't_ have `@batch_transform` (since, as you probably know from looking at the code, it is mostly for internal use), _but_ there have been some user-support in the past where I've realized it is useful!

And I've had to resort to the somewhat awkward

```python
@qml.batch_transform(qml.transforms.hamiltonian_expand)
@qml.qnode(dev)
```

to help the user 😆 

> then I have to somehow instantiate the tape of the qnode by executing the qnode to access the observables.

I'm not quite sure I follow what you mean here?

Reply:
A small code example:
```python
dev = qml.device("default.qubit", wires=1)

@qml.qnode(dev)
def circ():
    return qml.expval(qml.PauliZ(0))

print(circ.tape)
circ()
print(circ.tape)
```
returns:
```
None
<QuantumTape: wires=[0], params=0>
```

Reply:
Ah, you shouldn't need to worry about this when using the `@batch_transform` decorator 🙂 The function stays exactly the same: `tape` in, `tapes, fn` out. The decorator includes logic to ensure the transform automatically works with both tapes and QNodes, but you wouldn't need to worry about the QNode-specific logic

Reply:
What I meant was more about having `split_non_commuting()` as a function to act on qnodes, i.e. providing examples. Currently, something like `split_non_commuting(qnode)` doesnt work as it was written for tapes and obtains the observables via `tape.observables`. I could add a check if the input is a qnode then it can take it from `qnode.tape.observables` but this only works if the qnode has been executed once. Does that make sense?